internal/services/paymentdispatchers/circle_payment_dispatcher.go

```diff
@@ -224,36 +249,65 @@ func (c *CirclePaymentDispatcher) ensureRecipientIsReady(ctx context.Context, re
 	if err != nil {
 		return nil, fmt.Errorf("parsing Circle recipient status: %w", err)
 	}
-	updateDataRecipient := data.CircleRecipientUpdate{
+
+	dataRecipient, err = c.sdpModels.CircleRecipient.Update(ctx, dataRecipient.ReceiverWalletID, data.CircleRecipientUpdate{
 		IdempotencyKey:    dataRecipient.IdempotencyKey,
 		CircleRecipientID: recipient.ID,
 		Status:            dataRecipientStatus,
 		ResponseBody:      recipientJson,
+		SyncAttempts:      dataRecipient.SyncAttempts + 1,
+		LastSyncAttemptAt: time.Now(),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("updating Circle recipient: %w", err)
 	}
-	if shouldBumpSyncAttempts {
-		updateDataRecipient.SyncAttempts = dataRecipient.SyncAttempts + 1
-		updateDataRecipient.LastSyncAttemptAt = time.Now()
+
+	return dataRecipient, nil
+}
+
+// ensureRecipientIsReady ensures that the recipient is ready to receive payments, creating it in the database and in
+// the Circle API if needed.
+func (c *CirclePaymentDispatcher) ensureRecipientIsReady(ctx context.Context, receiverWallet data.ReceiverWallet) (*data.CircleRecipient, error) {
+	dataRecipient, err := c.getOrCreateRecipient(ctx, receiverWallet)
+	if err != nil {
+		return nil, fmt.Errorf("getting or creating Circle recipient: %w", err)
 	}
-	dataRecipient, err = c.sdpModels.CircleRecipient.Update(ctx, dataRecipient.ReceiverWalletID, updateDataRecipient)
+
+	// SUCCESS
+	if dataRecipient.Status == data.CircleRecipientStatusActive {
+		return dataRecipient, nil
+	}
+
+	// FAILED or INACTIVE
+	dataRecipient, err = c.handleFailedOrInactiveRecipientIfNeeded(ctx, dataRecipient)
 	if err != nil {
-		return nil, fmt.Errorf("updating Circle recipient: %w", err)
+		return nil, fmt.Errorf("handling failed or inactive recipient: %w", err)
+	}
+
+	// NULL, PENDING, INACTIVE (with renovated idempotency_key) or FAILED (with renovated idempotency_key) -> (re)submit the recipient creation request
+	dataRecipient, err = c.submitRecipientToCircle(ctx, receiverWallet, dataRecipient)
+	if err != nil {
+		return nil, fmt.Errorf("submitting recipient to Circle: %w", err)
 	}
 
 	return dataRecipient, nil
 }
 
-func (c *CirclePaymentDispatcher) ensureRecipientIsReadyWithRetry(ctx context.Context, receiverWallet data.ReceiverWallet) (*data.CircleRecipient, error) {
+// calls ensureRecipientIsReadyWithRetry with a retry policy.
+func (c *CirclePaymentDispatcher) ensureRecipientIsReadyWithRetry(ctx context.Context, receiverWallet data.ReceiverWallet, initialDelay time.Duration) (*data.CircleRecipient, error) {
+	if initialDelay <= 0 || initialDelay > time.Second {
+		initialDelay = 100 * time.Millisecond
+	}
+
 	var recipient *data.CircleRecipient
 	err := retry.Do(
 		func() error {
 			var err error
 			recipient, err = c.ensureRecipientIsReady(ctx, receiverWallet)
 			if err != nil {
 				if errors.Is(err, ErrCircleRecipientCreationFailedTooManyTimes) {
```

Comment:
❓I wonder if we will ever really reach this condition if we already have the `retry.Attempts(maxCircleRecipientCreationAttempts)`. 

If I understand this correctly : this is how a creation failure would go : 
1. Submit Circle Recipient 
2. Sync -> Recipient is Pending (retry 1)
3. Sync -> Recipient Error -> Try again (retry 2, SyncAttempts 1)
4. Sync -> Recipient Error -> Try again (retry 3, SyncAttempts 2)
5. Sync -> Recipient Error -> Try again (retry 4, SyncAttempts 3)
6. Sync -> Recipient Error -> Try again (retry 5, SyncAttempts 4) -> Abort. 

So maybe we should create two constants : 
* maxCircleRecipientCreationAttempts = 5
* maxCircleRecipientSyncAttempts > maxCircleRecipientCreationAttempts

Reply:
We could rely solely on the `ErrCircleRecipientCreationFailedTooManyTimes` being returned if we set the retry limit to `0` (considered unlimited retries by the lib).

The main difference would be:
- `retry.Attempts(maxCircleRecipientCreationAttempts)`: displays the aggregated errors from all failed (re)tries.
- `retry.Attempts(0)`: only displays the latest (re)tried error.

Reply:
Thank you for the explanation! Makes sense